Add Openai

HebeHH opened the pull request:
#  Add OpenAI

> Duplicated from #33 (Add OpenAI first PR)

## Reasoning

There's two major reasons to add additional providers to Lumentis:
1. It's always good to have additional providers as they have different strengths and user bases
2. Anthropic rate limits are killer and make Lumentis almost unusable to new users

## Methodology
This gives a wrapper `callLLM` function, which internally will use either `callAnthropic` or `callOpenAI`. 

### Wrapper
The `callLLM` wrapper is responsible for continuing on partial JSON if desired. This is not exactly relevant to `callOpenAI`, because OpenAI json mode doesn't allow for partial json in the response. However, structuring it in this way allows for us to add new providers which also need partial JSON continuance. 

The `callLLM` wrapper also handles saving messages and responses to files, since this should be done the same for all models.

### Specific functions
The specific provider calls handle:
* Adjusting for different JSON types
* Printing to console
* System prompts
* Counting tokens
* Actually calling the function

## Task list

### Completed

✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅  Add new apikey if different provider selected

### To do
⛔ JSON checks all passing
⛔ Continue if not enough JSON in the OpenAI models
⛔ Streaming
⛔ Testing and bundling

hrishioa commented on src/constants.ts:
```diff
@@ -11,27 +11,88 @@ export const WRITING_STYLE_SIZE_LIMIT = 10000;
+    tokenCountingModel: "gpt-3.5-turbo", // required bc OpenAI token counting is frustrating
+    smarterDescription:
+      "Pretty cheap, good at JSON, bit dumb for the management stuff",
+    pageDescription: "Fast, cheap, lower rate limit than Claude"
```
We probably mean 'better' rate limit than Claude

hrishioa commented on src/constants.ts:
```diff
@@ -11,27 +11,88 @@ export const WRITING_STYLE_SIZE_LIMIT = 10000;
+    tokenCountingModel: "gpt-4", // required bc OpenAI token counting is frustrating
+    smarterDescription: "Bit expensive but smart",
+    pageDescription:
+      "Good for high quality docs when you need to make a lot and are willing to pay"
```
We should probably add gpt-4o?

hrishioa commented on src/constants.ts:
```diff
@@ -11,27 +11,88 @@ export const WRITING_STYLE_SIZE_LIMIT = 10000;
+  "gpt-4-turbo-2024-04-09": {
+    provider: "openai",
+    tokenCountingModel: "gpt-4", // required bc OpenAI token counting is frustrating
+    inputTokensPerM: 10,
```
Jeez that's EX

HebeHH replied:
that's the cheaper one

HebeHH replied:
but pls do double check my pricing it would be embarassing to get that wrong

hrishioa replied:
dbl checked

hrishioa commented on src/app.ts:
```diff
@@ -195,19 +205,21 @@ async function runWizard() {
     (await password({
       message:
-        "Please enter an Anthropic API key.\n (You can leave this blank if it's already in the ENV variable.): ",
+        "Please enter your API key.\n (You can leave this blank if it's already in the ENV variable.): ",
```
We should probably ask with the specific provider name - less confusing to users.

hrishioa commented on src/app.ts:
```diff
@@ -219,12 +231,19 @@ async function runWizard() {
 
+  const baseOptions: AICallerOptions = {
+    model: wizardState.smarterModel,
+    maxOutputTokens: 700,
```
wai 700?

HebeHH replied:
...because that's what you were using?

hrishioa replied:
can't argue with myself- although I just pushed a commit contradicting me

hrishioa reviewed (commented):
Some comments - awesome PR!

hrishioa commented on src/ai.ts:
```diff
@@ -1,32 +1,219 @@
+    jsonType,
+    saveToFilepath,
+    prefix,
+    systemPrompt,
```
@HebeHH this seems unused?

HebeHH replied:
hmm it's used when the `options` are passed to the specific callProvider() functions, but I guess that doesn't mean they need to be pulled out here. 
Personally I kinda like being able to see what the parameters available to the function are, for future dev even if not in use right now, but it can also freely be removed.

hrishioa replied:
I think typescript means that you can see them from the typedef anyway :)

HebeHH replied:
yup but the typedef is a whole other file. However you can see it on hovering the typedef in VS Code so I shall remove

hrishioa commented on src/ai.ts:
```diff
@@ -48,71 +235,20 @@ export async function runClaudeInference(
+    let outputTokens = 0;
+    let inputTokens = 0;
+    if (AI_PROVIDERS[provider] === undefined) {
+      throw new Error("Invalid provider");
```
maybe this should instead be a check at the toplevel to make sure AI_MODELS_INFO has the model we're looking for?

HebeHH replied:
donesies

hrishioa commented:
Current TODOs:
1. Streaming
2. App-level cache for API keys instead of persistence
3. Adding Google to the providers for gemini-flash and pro
4. Checking source tokens and providing not just a Claude specific warning of too many tokens in source.
5. Generating multiple pages in parallel (since the ratelimits on gpt-4o are really quite high)

Which ones do you want me to pick up @HebeHH ?

HebeHH commented:
Have added `totalTokenLimit` and `outputTokenLimit` to the `AI_MODELS_INFO` recordset, and adjusted the `PRIMARY_SOURCE_BUDGET` to be a function that returns your primary source budget according to the specific model you are using.

Currently only the `totalTokenLimit`  is important - GPT has a lower one than Claude, so we need to take that into account when calculating the primary source budget.

HOWEVER @hrishioa do note: Gemini has twice the `outputTokenLimit` of currently added models. As such, when you add this, you might want to adjust the token limit for the larger calls (for the outline and the writing). You can either do this in `app.ts` by referencing `AI_MODELS_INFO`, or just pass `options.maxOutputTokens = 999999999`. I've added to `callLLM()` : 
```
const maxOutputTokens = Math.min(
    AI_MODELS_INFO[model].outputTokenLimit, 
    options.maxOutputTokens
  )
  ```
  
  So it will just reset to the max output token of the model automatically.

hrishioa commented:
Streaming done, tested and pushed!

HebeHH commented:
Completed
✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅ Add new apikey if different provider selected
✅ JSON checks all passing
✅ Streaming

To do
⛔ Continue if not enough JSON in the OpenAI models
⛔ Testing and bundling

_Continue if not enough JSON notes:_
> The JSON in the message the model returns may be partial (i.e. cut off) if finish_reason is length, which indicates the generation exceeded max_tokens or the conversation exceeded the token limit. To guard against this, check finish_reason before parsing the response. _- [OpenAI website](https://platform.openai.com/docs/guides/text-generation/json-mode)_

HebeHH commented:
Completed
✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅ Add new apikey if different provider selected
✅ JSON checks all passing
✅ Streaming
✅ Checking source tokens and providing not just a Claude specific warning of too many tokens in source.

To do
⛔ Continue if not enough JSON in the OpenAI models
⛔ Testing and bundling
⛔ App-level cache for API keys instead of persistence
⛔ Adding Google to the providers for gemini-flash and pro
⛔ Generating multiple pages in parallel (since the ratelimits on gpt-4o are really quite high)
⛔ Testing existing functionality with full package and different platforms

hrishioa commented:
Completed
✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅ Add new apikey if different provider selected
✅ JSON checks all passing
✅ Streaming
✅ Checking source tokens and providing not just a Claude specific warning of too many tokens in source.
✅ Generating multiple pages in parallel (since the ratelimits on gpt-4o are really quite high)


To do
⛔ Continue if not enough JSON in the OpenAI models
⛔ Testing and bundling
⛔ Adding Google to the providers for gemini-flash and pro
⛔ App-level cache for API keys instead of persistence
⛔ Testing existing functionality with full package and different platforms

HebeHH commented:
**Completed**
✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅ Add new apikey if different provider selected
✅ JSON checks all passing
✅ Don't save API key in plaintext file
✅ Streaming
✅ Checking source tokens and providing not just a Claude specific warning of too many tokens in source.
✅ Generating multiple pages in parallel (since the ratelimits on gpt-4o are really quite high)

**In Progress**
🧿 Testing

**To do**
⛔ Continue if not enough JSON in the OpenAI models
⛔ Bundling (and testing thereof)
⛔ Adding Google to the providers for gemini-flash and pro
⛔ Testing existing functionality with full package and different platforms

**Cancelled**
🤷‍♀️ App-level cache for API keys instead of persistence. _Decided after discussion that the current way of handling apikeys is fine._

HebeHH commented:
**Completed**
✅ Added OpenAI calling functionality (with json forcing)
✅ Moved option parameters to a dictionary
✅ Token and cost counting appropriate to different models
✅ Select OpenAI models from the Wizard
✅ Select different models
✅ Extensible handling of additional providers
✅ Add new apikey if different provider selected
✅ JSON checks all passing
✅ Don't save API key in plaintext file
✅ Streaming
✅ Checking source tokens and providing not just a Claude specific warning of too many tokens in source.
✅ Generating multiple pages in parallel (since the ratelimits on gpt-4o are really quite high)
✅ Adding Google to the providers for gemini-flash and pro
✅ Bundling (and testing thereof)
✅ Handle incomplete JSON in `json_mode`

**In Progress**
🧿 Testing

**To do**
⛔ Continue if not enough JSON in the OpenAI models
⛔ Testing existing functionality with full package and different platforms
⛔ Update README

**Cancelled**
🤷‍♀️ App-level cache for API keys instead of persistence. _Decided after discussion that the current way of handling apikeys is fine._

**Discussion: Continue on not enough JSON. **
This is a problem because the `json_mode` models (OpenAI, Gemini) are _wonderful_ for giving complete JSON, but terrible if they run out of context length. They'll just cut out halfway through and give incomplete JSON. 

Unfortunately the existing 'continue on partial JSON' functionality doesn't work with this, because a) these models don't let you put words in their mouth (can't continue assistant prompt) b) a `json_mode` continuance will start a new object instead of trying to complete the existing one.

Methods in consideration:
1. Just ask the model for a shorter object
2. Just retry
3. Use `partialParse()` to forcefully close the object and then just use that
4. Use `partialParse()`, remove the last toplevel object, and then ask the model to continue making more objects, then merge them

I have currently implemented (3) but invite @hrishioa to try implementing (4) as it is probably better

hrishioa reviewed (approved):
Good to go! Final testing in progress.
